README.md
# Logback awslogs JSON encoder

[![build](https://github.com/osiegmar/logback-awslogs-json-encoder/actions/workflows/build.yml/badge.svg)](https://github.com/osiegmar/logback-awslogs-json-encoder/actions/workflows/build.yml)
[![javadoc](https://javadoc.io/badge2/de.siegmar/logback-awslogs-json-encoder/javadoc.svg)](https://javadoc.io/doc/de.siegmar/logback-awslogs-json-encoder)
[![Maven Central](https://img.shields.io/maven-central/v/de.siegmar/logback-awslogs-json-encoder.svg)](https://central.sonatype.com/artifact/de.siegmar/logback-awslogs-json-encoder)

Logback encoder for producing JSON output that is handled by
AWS [CloudWatch Logs Insights](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AnalyzingLogData.html). This
library has no external dependencies and thus very light footprint.

Since version 1.3.8, Logback ships with a [JsonEncoder](https://logback.qos.ch/manual/encoders.html#JsonEncoder) itself.
Unfortunately that JsonEncoder produces log output that is poorly suited for use with CloudWatch Logs (Insights).

## Features

- Forwarding of MDC (Mapped Diagnostic Context)
- Forwarding of caller data
- Forwarding of static fields
- Forwarding of exception root cause
- No runtime dependencies beside Logback

## Requirements

- Java 11
- Logback 1.4.11

## Prerequisites

Ensure that the task definition of your ECS task uses the `awslogs` log driver.

A full example (excerpt from a full task definition JSON) could look like this:

```json
{
  "logConfiguration": {
    "logDriver": "awslogs",
    "options": {
      "awslogs-group": "myloggroup",
      "awslogs-region": "eu-central-1"
    }
  }
}
```

## Configuration

A basic configuration with defaults would look like this:

```xml
<configuration>

    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="de.siegmar.logbackawslogsjsonencoder.AwsJsonLogEncoder"/>
    </appender>

    <root level="DEBUG">
        <appender-ref ref="STDOUT"/>
    </root>

</configuration>
```

An extended version (with all default values set explicitly, custom static fields and mapper) would look like this:

```xml
<configuration>

  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
    <encoder class="de.siegmar.logbackawslogsjsonencoder.AwsJsonLogEncoder">
      <includeTimestamp>true</includeTimestamp>
      <includeNanoseconds>false</includeNanoseconds>
      <includeSequenceNumber>false</includeSequenceNumber>
      <includeLevelName>true</includeLevelName>
      <includeThreadName>true</includeThreadName>
      <includeLoggerName>true</includeLoggerName>
      <includeFormattedMessage>true</includeFormattedMessage>
      <includeRawMessage>false</includeRawMessage>
      <includeStacktrace>true</includeStacktrace>
      <includeRootCause>false</includeRootCause>
      <includeMarker>true</includeMarker>
      <includeMdc>true</includeMdc>
      <includeKeyValues>true</includeKeyValues>
      <includeCaller>false</includeCaller>

      <staticField>app_name:backend</staticField>
      <staticField>os_arch:${os.arch}</staticField>

      <customMapper class="your.custom.Mapper"/>
    </encoder>
  </appender>

  <root level="DEBUG">
    <appender-ref ref="STDOUT"/>
  </root>

</configuration>
```

## Example output

Typical output:

```json
{
  "timestamp": 1698595093642,
  "level": "DEBUG",
  "thread": "Test worker",
  "logger": "my.app.MyClass",
  "message": "Message 1"
}
```

Maximal output:

```json
{
  "timestamp": 1698595093642,
  "nanoseconds": 642725000,
  "sequenceNumber": 0,
  "level": "DEBUG",
  "thread": "Test worker",
  "logger": "my.app.MyClass",
  "message": "Message 1",
  "rawMessage": "Message {}",
  "markers": {
    "foo": 1,
    "bar": 1
  },
  "mdc": {
    "foo": "bar",
    "baz": null
  },
  "keyValues": {
    "foo": "bar",
    "bar": null,
    "null": "bar"
  },
  "caller": {
    "file": "NativeMethodAccessorImpl.java",
    "line": -2,
    "class": "jdk.internal.reflect.NativeMethodAccessorImpl",
    "method": "invoke0"
  },
  "stacktrace": "java.lang.IllegalStateException: Error processing data\n\tat my.app.MyClass.foo(MyClass.java:123)",
  "rootCause": {
    "class": "java.lang.IllegalStateException",
    "message": "Error processing data"
  },
  "staticFields": {
    "app_name": "backend",
    "os_arch": "amd64"
  },
  "custom": "A field added by a custom mapper"
}
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.1] - 2023-12-10
### Changed
- Updated Logback

## [2.0.0] - 2023-11-05
### Changed
- Complete rewrite
- Update to Java 11
- Update Logback dependency to 1.4.11

## [1.1.1] - 2022-01-04
### Changed
- Update Logback dependency to 1.2.10

## [1.1.0] - 2020-08-23
### Added
- Add support for MDC null values
  [\#2](https://github.com/osiegmar/logback-awslogs-json-encoder/pull/2)

## [1.0.0] - 2018-12-09
- Initial release

[2.0.1]: https://github.com/osiegmar/logback-awslogs-json-encoder/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/osiegmar/logback-awslogs-json-encoder/compare/v1.1.1...v2.0.0
[1.1.1]: https://github.com/osiegmar/logback-awslogs-json-encoder/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/osiegmar/logback-awslogs-json-encoder/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/osiegmar/logback-awslogs-json-encoder/releases/tag/v1.0.0


src/test/java/de/siegmar/logbackawslogsjsonencoder/MyCustomMapper.java
/*
 * Logback awslogs JSON encoder.
 * Copyright (C) 2023 Oliver Siegmar
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package de.siegmar.logbackawslogsjsonencoder;

import java.util.function.BiConsumer;

import ch.qos.logback.classic.spi.ILoggingEvent;

public class MyCustomMapper implements BiConsumer<SimpleJsonEncoder, ILoggingEvent> {

    @Override
    public void accept(final SimpleJsonEncoder json, final ILoggingEvent event) {
        json.append("custom", "bar");
    }

}


src/test/java/de/siegmar/logbackawslogsjsonencoder/AwsJsonLogEncoderTest.java
/*
 * Logback awslogs JSON encoder.
 * Copyright (C) 2023 Oliver Siegmar
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package de.siegmar.logbackawslogsjsonencoder;

import static net.javacrumbs.jsonunit.assertj.JsonAssertions.assertThatJson;
import static net.javacrumbs.jsonunit.assertj.JsonAssertions.json;

import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

import org.junit.jupiter.api.Test;
import org.slf4j.LoggerFactory;
import org.slf4j.MarkerFactory;
import org.slf4j.event.KeyValuePair;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.LoggerContext;
import ch.qos.logback.classic.spi.LoggingEvent;

@SuppressWarnings("PMD.AvoidDuplicateLiterals")
class AwsJsonLogEncoderTest {

    private static final String LOGGER_NAME = AwsJsonLogEncoderTest.class.getCanonicalName();

    private final AwsJsonLogEncoder encoder = new AwsJsonLogEncoder();

    AwsJsonLogEncoderTest() {
        encoder.setContext(new LoggerContext());
    }

    @Test
    void defaultConfig() {
        encoder.start();

        //language=JSON5
        final String expectedJson =
            "{"
            + "timestamp: '${json-unit.any-number}',"
            + "level: 'DEBUG',"
            + "thread: 'Test worker',"
            + "logger: 'de.siegmar.logbackawslogsjsonencoder.AwsJsonLogEncoderTest',"
            + "message: 'message 1',"
            + "markers: {"
            + "    'foo': 1,"
            + "    'bar': 1"
            + "},"
            + "mdc: {"
            + "    foo: 'bar',"
            + "    baz: null"
            + "},"
            + "keyValues: {"
            + "    foo: 'bar',"
            + "    bar: null,"
            + "    'null': 'bar'"
            + "}}";
        assertThatJson(fullLog()).isEqualTo(json(expectedJson));
    }

    @Test
    void disableAll() {
        setupAllDisabledEncoder(c -> {
        });
        assertThatJson(fullLog()).isEqualTo("{}");
    }

    @Test
    void timestamp() {
        setupAllDisabledEncoder(c -> c.setIncludeTimestamp(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'timestamp': '${json-unit.any-number}'}"));
    }

    @Test
    void nanoseconds() {
        setupAllDisabledEncoder(c -> c.setIncludeNanoseconds(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'nanoseconds': '${json-unit.any-number}'}"));
    }

    @Test
    void sequenceNumber() {
        setupAllDisabledEncoder(c -> c.setIncludeSequenceNumber(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'sequenceNumber': '${json-unit.any-number}'}"));
    }

    @Test
    void levelName() {
        setupAllDisabledEncoder(c -> c.setIncludeLevelName(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'level': 'DEBUG'}"));
    }

    @Test
    void threadName() {
        setupAllDisabledEncoder(c -> c.setIncludeThreadName(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'thread': 'Test worker'}"));
    }

    @Test
    void loggerName() {
        setupAllDisabledEncoder(c -> c.setIncludeLoggerName(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'logger': 'de.siegmar.logbackawslogsjsonencoder.AwsJsonLogEncoderTest'}"));
    }

    @Test
    void formattedMessage() {
        setupAllDisabledEncoder(c -> c.setIncludeFormattedMessage(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'message': 'message 1'}"));
    }

    @Test
    void rawMessage() {
        setupAllDisabledEncoder(c -> c.setIncludeRawMessage(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'rawMessage': 'message {}'}"));
    }

    @Test
    void stacktrace() {
        setupAllDisabledEncoder(c -> c.setIncludeStacktrace(true));

        final String logMsg;
        try {
            throw new IllegalArgumentException("Example Exception");
        } catch (final IllegalArgumentException e) {
            logMsg = fullLog(e);
        }

        assertThatJson(logMsg).and(
            j -> j.node("stacktrace").isString().startsWith(
                "java.lang.IllegalArgumentException: Example Exception\n"
                + "\tat de.siegmar.logbackawslogsjsonencoder.AwsJsonLogEncoderTest.stacktrace(AwsJsonLogEncoderTest")
        );
    }

    @SuppressWarnings("checkstyle:IllegalCatch")
    @Test
    void rootCause() {
        setupAllDisabledEncoder(c -> c.setIncludeRootCause(true));

        final String logMsg;
        try {
            throw new RuntimeException(new IllegalArgumentException("Example Exception"));
        } catch (final RuntimeException e) {
            logMsg = fullLog(e);
        }

        //language=JSON5
        final String expectedJson =
            "{rootCause: {"
            + "class: 'java.lang.IllegalArgumentException',"
            + "message: 'Example Exception'"
            + "}}";

        assertThatJson(logMsg).isEqualTo(json(expectedJson));
    }

    @Test
    void marker() {
        setupAllDisabledEncoder(c -> c.setIncludeMarker(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'markers': {'foo': 1, 'bar': 1}}"));
    }

    @Test
    void mdc() {
        setupAllDisabledEncoder(c -> c.setIncludeMdc(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'mdc': {'foo': 'bar', 'baz': null}}"));
    }

    @Test
    void keyValues() {
        setupAllDisabledEncoder(c -> c.setIncludeKeyValues(true));
        assertThatJson(fullLog())
            .isEqualTo(json("{'keyValues': {'foo': 'bar', 'bar': null, 'null': 'bar'}}"));
    }

    @Test
    void caller() {
        setupAllDisabledEncoder(c -> c.setIncludeCaller(true));
        //language=JSON5
        final String expectedJson =
            "{caller: {"
            + "    file: 'NativeMethodAccessorImpl.java',"
            + "    line: -2,"
            + "    class: 'jdk.internal.reflect.NativeMethodAccessorImpl',"
            + "    method: 'invoke0'"
            + "}}";
        assertThatJson(fullLog())
            .isEqualTo(json(expectedJson));
    }

    @Test
    void staticFields() {
        setupAllDisabledEncoder(c -> c.addStaticField("foo:bar"));
        assertThatJson(fullLog())
            .isEqualTo(json("{'staticFields': {'foo': 'bar'}}"));
    }

    @Test
    void customMapper() {
        setupAllDisabledEncoder(c -> c.addCustomMapper(new MyCustomMapper()));
        assertThatJson(fullLog())
            .isEqualTo(json("{'custom': 'bar'}"));
    }

    @Test
    void complex() {
        setupAllEnabledEncoder(c -> c.addStaticField("foo:bar"));

        //language=JSON5
        final String expectedJson =
            "{"
            + "    sequenceNumber: 0,"
            + "    timestamp: '${json-unit.any-number}',"
            + "    nanoseconds: '${json-unit.any-number}',"
            + "    level: 'DEBUG',"
            + "    thread: 'Test worker',"
            + "    logger: 'de.siegmar.logbackawslogsjsonencoder.AwsJsonLogEncoderTest',"
            + "    message: 'message 1',"
            + "    rawMessage: 'message {}',"
            + "    staticFields: {"
            + "        foo: 'bar'"
            + "    },"
            + "    markers: {"
            + "        foo: 1,"
            + "        bar: 1"
            + "    },"
            + "    mdc: {"
            + "        foo: 'bar',"
            + "        baz: null"
            + "    },"
            + "    keyValues: {"
            + "        foo: 'bar',"
            + "        bar: null,"
            + "        'null': 'bar'"
            + "    },"
            + "    caller: {"
            + "        file: 'NativeMethodAccessorImpl.java',"
            + "        line: -2,"
            + "        class: 'jdk.internal.reflect.NativeMethodAccessorImpl',"
            + "        method: 'invoke0'"
            + "    }"
            + "}";

        assertThatJson(fullLog()).isEqualTo(json(expectedJson));
    }

    private void setupAllDisabledEncoder(final Consumer<AwsJsonLogEncoder> customize) {
        encoder.setIncludeTimestamp(false);
        encoder.setIncludeNanoseconds(false);
        encoder.setIncludeSequenceNumber(false);
        encoder.setIncludeLevelName(false);
        encoder.setIncludeThreadName(false);
        encoder.setIncludeLoggerName(false);
        encoder.setIncludeFormattedMessage(false);
        encoder.setIncludeRawMessage(false);
        encoder.setIncludeStacktrace(false);
        encoder.setIncludeRootCause(false);
        encoder.setIncludeMarker(false);
        encoder.setIncludeMdc(false);
        encoder.setIncludeKeyValues(false);
        encoder.setIncludeCaller(false);
        customize.accept(encoder);
        encoder.start();
    }

    private void setupAllEnabledEncoder(final Consumer<AwsJsonLogEncoder> customize) {
        encoder.setIncludeTimestamp(true);
        encoder.setIncludeNanoseconds(true);
        encoder.setIncludeSequenceNumber(true);
        encoder.setIncludeLevelName(true);
        encoder.setIncludeThreadName(true);
        encoder.setIncludeLoggerName(true);
        encoder.setIncludeFormattedMessage(true);
        encoder.setIncludeRawMessage(true);
        encoder.setIncludeStacktrace(true);
        encoder.setIncludeRootCause(true);
        encoder.setIncludeMarker(true);
        encoder.setIncludeMdc(true);
        encoder.setIncludeKeyValues(true);
        encoder.setIncludeCaller(true);
        customize.accept(encoder);
        encoder.start();
    }

    private String fullLog() {
        return fullLog(null);
    }

    private String fullLog(final Exception e) {
        return dummyLog(c -> {
            c.addMarker(MarkerFactory.getMarker("foo"));
            c.addMarker(MarkerFactory.getMarker("bar"));

            final Map<String, String> mdcMap = new LinkedHashMap<>();
            mdcMap.put("foo", "bar");
            mdcMap.put("baz", null);
            c.setMDCPropertyMap(mdcMap);

            c.setKeyValuePairs(List.of(
                new KeyValuePair("foo", "bar"),
                new KeyValuePair("bar", null),
                new KeyValuePair(null, "bar")
            ));
        }, e);
    }

    private String dummyLog(final Consumer<LoggingEvent> customize, final Throwable e) {
        final LoggerContext lc = (LoggerContext) LoggerFactory.getILoggerFactory();
        final Logger logger = lc.getLogger(LOGGER_NAME);
        final LoggingEvent event = new LoggingEvent(
            LOGGER_NAME,
            logger,
            Level.DEBUG,
            "message {}",
            e,
            new Object[]{1});

        customize.accept(event);
        return new String(encoder.encode(event), StandardCharsets.UTF_8);
    }

}


src/test/java/de/siegmar/logbackawslogsjsonencoder/SimpleJsonEncoderTest.java
/*
 * Logback awslogs JSON encoder.
 * Copyright (C) 2023 Oliver Siegmar
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package de.siegmar.logbackawslogsjsonencoder;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

@SuppressWarnings({"PMD.AvoidStringBufferField", "PMD.AvoidDuplicateLiterals"})
class SimpleJsonEncoderTest {

    private final StringBuilder sb = new StringBuilder();
    private final SimpleJsonEncoder enc = new SimpleJsonEncoder(sb);

    @Test
    void string() {
        enc.append("aaa", "bbb");
        assertThat(produce()).isEqualTo("{\"aaa\":\"bbb\"}");
    }

    private String produce() {
        enc.end();
        return sb.toString();
    }

    @Test
    void number() {
        enc.append("aaa", 123);
        assertThat(produce()).isEqualTo("{\"aaa\":123}");
    }

    @Test
    void quote() {
        enc.append("aaa", "\"");
        assertThat(produce()).isEqualTo("{\"aaa\":\"\\\"\"}");
    }

    @Test
    void reverseSolidus() {
        enc.append("aaa", "\\");
        assertThat(produce()).isEqualTo("{\"aaa\":\"\\\\\"}");
    }

    @Test
    void solidus() {
        enc.append("aaa", "/");
        assertThat(produce()).isEqualTo("{\"aaa\":\"\\/\"}");
    }

    @Test
    void backspace() {
        enc.append("aaa", "\b");
        assertThat(produce()).isEqualTo("{\"aaa\":\"\\b\"}");
    }

    @Test
    void formFeed() {
        enc.append("aaa", "\f");
        assertThat(produce()).isEqualTo("{\"aaa\":\"\\f\"}");
    }

    @Test
    void newline() {
        enc.append("aaa", "\n");
        assertThat(produce()).isEqualTo("{\"aaa\":\"\\n\"}");
    }

    @Test
    void carriageReturn() {
        enc.append("aaa", "\r");
        assertThat(produce()).isEqualTo("{\"aaa\":\"\\r\"}");
    }

    @Test
    void tab() {
        enc.append("aaa", "\t");
        assertThat(produce()).isEqualTo("{\"aaa\":\"\\t\"}");
    }

    @Test
    @SuppressWarnings("checkstyle:avoidescapedunicodecharacters")
    void unicode() {
        enc.append("\u0002", "\u0007\u0019");
        assertThat(produce()).isEqualTo("{\"\\u0002\":\"\\u0007\\u0019\"}");
    }

    @Test
    void multipleFields() {
        enc.append("bbb", "ccc");
        enc.append("ddd", 123);

        assertThat(produce()).isEqualTo("{\"bbb\":\"ccc\",\"ddd\":123}");
    }

}


src/main/java/de/siegmar/logbackawslogsjsonencoder/AwsJsonLogEncoder.java
/*
 * Logback awslogs JSON encoder.
 * Copyright (C) 2023 Oliver Siegmar
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package de.siegmar.logbackawslogsjsonencoder;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.BiConsumer;

import org.slf4j.Marker;
import org.slf4j.event.KeyValuePair;

import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.classic.spi.IThrowableProxy;
import ch.qos.logback.classic.spi.ThrowableProxyUtil;
import ch.qos.logback.core.encoder.EncoderBase;

/**
 * Logback encoder that produces JSON that is read by CloudWatch Logs Insights.
 */
public class AwsJsonLogEncoder extends EncoderBase<ILoggingEvent> {

    private static final int INITIAL_BUFFER_SIZE = 256;

    private final Map<String, Object> staticFields = new LinkedHashMap<>();
    private final List<BiConsumer<SimpleJsonEncoder, ILoggingEvent>> mappers = new ArrayList<>();
    private final List<BiConsumer<SimpleJsonEncoder, ILoggingEvent>> customMappers = new ArrayList<>();

    private boolean includeTimestamp = true;
    private boolean includeNanoseconds;
    private boolean includeSequenceNumber;
    private boolean includeLevelName = true;
    private boolean includeThreadName = true;
    private boolean includeLoggerName = true;
    private boolean includeFormattedMessage = true;
    private boolean includeRawMessage;
    private boolean includeStacktrace = true;
    private boolean includeRootCause;
    private boolean includeMarker = true;
    private boolean includeMdc = true;
    private boolean includeKeyValues = true;
    private boolean includeCaller;

    public Map<String, Object> getStaticFields() {
        return staticFields;
    }

    @SuppressWarnings("PMD.AvoidLiteralsInIfCondition")
    public void addStaticField(final String staticField) {
        final String[] split = staticField.split(":", 2);
        if (split.length == 2) {
            addField(staticFields, split[0].trim(), split[1].trim());
        } else {
            addWarn("staticField must be in format key:value - rejecting '" + staticField + "'");
        }
    }

    public List<BiConsumer<SimpleJsonEncoder, ILoggingEvent>> getCustomMappers() {
        return customMappers;
    }

    public void addCustomMapper(final BiConsumer<SimpleJsonEncoder, ILoggingEvent> customMapper) {
        customMappers.add(customMapper);
    }

    public boolean isIncludeTimestamp() {
        return includeTimestamp;
    }

    public void setIncludeTimestamp(final boolean includeTimestamp) {
        this.includeTimestamp = includeTimestamp;
    }

    public boolean isIncludeNanoseconds() {
        return includeNanoseconds;
    }

    public void setIncludeNanoseconds(final boolean includeNanoseconds) {
        this.includeNanoseconds = includeNanoseconds;
    }

    public boolean isIncludeSequenceNumber() {
        return includeSequenceNumber;
    }

    public void setIncludeSequenceNumber(final boolean includeSequenceNumber) {
        this.includeSequenceNumber = includeSequenceNumber;
    }

    public boolean isIncludeLevelName() {
        return includeLevelName;
    }

    public void setIncludeLevelName(final boolean includeLevelName) {
        this.includeLevelName = includeLevelName;
    }

    public boolean isIncludeThreadName() {
        return includeThreadName;
    }

    public void setIncludeThreadName(final boolean includeThreadName) {
        this.includeThreadName = includeThreadName;
    }

    public boolean isIncludeLoggerName() {
        return includeLoggerName;
    }

    public void setIncludeLoggerName(final boolean includeLoggerName) {
        this.includeLoggerName = includeLoggerName;
    }

    public boolean isIncludeFormattedMessage() {
        return includeFormattedMessage;
    }

    public void setIncludeFormattedMessage(final boolean includeFormattedMessage) {
        this.includeFormattedMessage = includeFormattedMessage;
    }

    public boolean isIncludeRawMessage() {
        return includeRawMessage;
    }

    public void setIncludeRawMessage(final boolean includeRawMessage) {
        this.includeRawMessage = includeRawMessage;
    }

    public boolean isIncludeStacktrace() {
        return includeStacktrace;
    }

    public void setIncludeStacktrace(final boolean includeStacktrace) {
        this.includeStacktrace = includeStacktrace;
    }

    public boolean isIncludeRootCause() {
        return includeRootCause;
    }

    public void setIncludeRootCause(final boolean includeRootCause) {
        this.includeRootCause = includeRootCause;
    }

    public boolean isIncludeMarker() {
        return includeMarker;
    }

    public void setIncludeMarker(final boolean includeMarker) {
        this.includeMarker = includeMarker;
    }

    public boolean isIncludeMdc() {
        return includeMdc;
    }

    public void setIncludeMdc(final boolean includeMdc) {
        this.includeMdc = includeMdc;
    }

    public boolean isIncludeKeyValues() {
        return includeKeyValues;
    }

    public void setIncludeKeyValues(final boolean includeKeyValues) {
        this.includeKeyValues = includeKeyValues;
    }

    public boolean isIncludeCaller() {
        return includeCaller;
    }

    public void setIncludeCaller(final boolean includeCaller) {
        this.includeCaller = includeCaller;
    }

    private void addField(final Map<String, Object> dst, final String key, final String value) {
        if (key.isEmpty()) {
            addWarn("staticField key must not be empty");
        } else if (dst.containsKey(key)) {
            addWarn("additional field with key '" + key + "' is already set");
        } else {
            dst.put(key, value);
        }
    }

    @SuppressWarnings({"checkstyle:NPathComplexity", "checkstyle:CyclomaticComplexity"})
    @Override
    public void start() {
        if (includeTimestamp) {
            mappers.add((json, event) -> json.append("timestamp", event.getTimeStamp()));
        }
        if (includeNanoseconds) {
            mappers.add((json, event) -> json.append("nanoseconds", event.getNanoseconds()));
        }
        if (includeSequenceNumber) {
            mappers.add((json, event) -> json.append("sequenceNumber", event.getSequenceNumber()));
        }
        if (includeLevelName) {
            mappers.add((json, event) -> json.append("level", event.getLevel().toString()));
        }
        if (includeThreadName) {
            mappers.add((json, event) -> json.append("thread", event.getThreadName()));
        }
        if (includeLoggerName) {
            mappers.add((json, event) -> json.append("logger", event.getLoggerName()));
        }
        if (includeFormattedMessage) {
            mappers.add((json, event) -> json.append("message", event.getFormattedMessage()));
        }
        if (includeRawMessage) {
            mappers.add((json, event) -> json.append("rawMessage", event.getMessage()));
        }
        if (includeMarker) {
            mappers.add((json, event) -> appendMarker(json, event.getMarkerList()));
        }
        if (includeMdc) {
            mappers.add((json, event) -> appendMdc(json, event.getMDCPropertyMap()));
        }
        if (includeKeyValues) {
            mappers.add((json, event) -> appendKeyValues(json, event.getKeyValuePairs()));
        }
        if (includeCaller) {
            mappers.add((json, event) -> appendCaller(json, event.getCallerData()));
        }
        if (includeStacktrace) {
            mappers.add((json, event) -> appendThrowable(json, event.getThrowableProxy()));
        }
        if (includeRootCause) {
            mappers.add((json, event) -> appendRootCause(json, event.getThrowableProxy()));
        }
        if (!staticFields.isEmpty()) {
            mappers.add((json, event) -> appendStaticFields(json, staticFields));
        }

        mappers.addAll(customMappers);

        super.start();
    }

    @Override
    public byte[] encode(final ILoggingEvent event) {
        final StringBuilder sb = new StringBuilder(INITIAL_BUFFER_SIZE);

        final var json = new SimpleJsonEncoder(sb);
        mappers.forEach(m -> m.accept(json, event));
        json.end();

        sb.append(System.lineSeparator());

        return sb.toString().getBytes(StandardCharsets.UTF_8);
    }

    private static void appendMarker(final SimpleJsonEncoder json, final List<Marker> markerList) {
        if (markerList == null || markerList.isEmpty()) {
            return;
        }

        json.appendObject("markers", j ->
            markerList.forEach(marker -> j.append(marker.getName(), 1)));
    }

    private static void appendMdc(final SimpleJsonEncoder json, final Map<String, String> mdcProperties) {
        if (mdcProperties == null || mdcProperties.isEmpty()) {
            return;
        }

        json.appendObject("mdc", j ->
            mdcProperties.forEach(j::append));
    }

    private static void appendKeyValues(final SimpleJsonEncoder json, final List<KeyValuePair> keyValuePairs) {
        if (keyValuePairs == null || keyValuePairs.isEmpty()) {
            return;
        }

        json.appendObject("keyValues", j ->
            keyValuePairs.forEach(kvp -> j.append(kvp.key, kvp.value)));
    }

    @SuppressWarnings("PMD.UseVarargs")
    private static void appendCaller(final SimpleJsonEncoder json, final StackTraceElement[] stackTraceElements) {
        if (stackTraceElements == null || stackTraceElements.length == 0) {
            return;
        }

        final StackTraceElement first = stackTraceElements[0];
        json.appendObject("caller", j -> j
            .append("file", first.getFileName())
            .append("line", first.getLineNumber())
            .append("class", first.getClassName())
            .append("method", first.getMethodName()));
    }

    private static void appendThrowable(final SimpleJsonEncoder json, final IThrowableProxy throwableProxy) {
        if (throwableProxy == null) {
            return;
        }

        json.append("stacktrace", ThrowableProxyUtil.asString(throwableProxy));
    }

    private static void appendRootCause(final SimpleJsonEncoder json, final IThrowableProxy throwableProxy) {
        findRootException(throwableProxy).ifPresent(rootException ->
            json.appendObject("rootCause", j -> j
                .append("class", rootException.getClassName())
                .append("message", rootException.getMessage())));
    }

    private static Optional<IThrowableProxy> findRootException(final IThrowableProxy throwableProxy) {
        if (throwableProxy == null) {
            return Optional.empty();
        }

        IThrowableProxy rootCause = throwableProxy;
        while (rootCause.getCause() != null) {
            rootCause = rootCause.getCause();
        }

        return Optional.of(rootCause);
    }

    private static void appendStaticFields(final SimpleJsonEncoder json, final Map<String, Object> staticFields) {
        json.appendObject("staticFields", j ->
            staticFields.forEach(j::append));
    }

    @SuppressWarnings("PMD.ReturnEmptyCollectionRatherThanNull")
    @Override
    public byte[] headerBytes() {
        return null;
    }

    @SuppressWarnings("PMD.ReturnEmptyCollectionRatherThanNull")
    @Override
    public byte[] footerBytes() {
        return null;
    }

}
